---
title: "BCoAT analyses"
title-block-banner: true
subtitle: "Analyses of the BCoAT viral contigs"
author: 
  - name: "Jinling Xue"
    orcid: "0000-0002-8191-4140"
date: 2022-09-11
format:
  html:
    toc: true
    code-fold: true
    anchor-sections: true
    smooth-scrol: true
    link-external-newwindow: true
    page-layout: full
bibliography: refs.bib
params:
  name: "20220911-BCoAT"
---

Updated: `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'CET')` CET.

```{r setup, message=FALSE, include=F}
here::i_am(paste0(params$name, ".qmd"), uuid = "389d9fd1-55a3-490a-8e51-e1d734b8625c")
projthis::proj_create_dir_target(params$name, clean = F)
path_target <- projthis::proj_path_target(params$name)
path_source <- projthis::proj_path_source(params$name)
```

The purpose of this document is to describe the analyses of the BCoAT viral contigs.

# Contigs contain BCoAT gene

```{r setup2}
#| echo: false
#| warning: false
library(here)
library(tidyverse)
library(data.table)

# file path
prj_path <- normalizePath("..")
data_path <- here(prj_path, "hpc/hms/virmerge/files")
fp_vc <- here(data_path, "clusters/genome_by_genome_overview.csv")

# functions
get_vc_members_by_member <- function(vc, member_id) {
  # vc: genome_by_genome_overview.csv table
  # member_id: Genome id in vc
  vc_id = as.character(vc[vc$Genome == member_id,"VC"])
  
  vc_members = vc %>% 
    dplyr::filter(VC == vc_id) %>% 
    dplyr::select(c(Genome, VC))
  
  return(vc_members)
}


# import data
vc <- fread(fp_vc) %>% dplyr::select(-V1)
```

```{r contigs_contain_BCoAT}
# Contigs contain BCoAT
contigs_contain_bcoat <- c("BAM6_V__NODE_3_length_90756_cov_3.612584-cat_2",     # Bin: BAM6_V__183
                           "BAF3_V__NODE_17_length_77772_cov_3.675859-cat_2")    # Bin: BAF3_V__502

```

These contigs are in bin:

-   BAM6_V\_\_183: BAM6_V\_\_NODE_3\_length_90756_cov_3.612584-cat_2
-   BAF3_V\_\_502: BAF3_V\_\_NODE_17_length_77772_cov_3.675859-cat_2

```{r get_VC_members}
#| echo: false
#| warning: false
# function get_vc_members_by_member apply to contigs_contain_bcoat
vc_members <- lapply(contigs_contain_bcoat, get_vc_members_by_member, vc = vc)
names(vc_members) <- contigs_contain_bcoat

vc_members <- vc_members %>% 
  purrr::map_df(~ .x, .id = "representative_contig") %>% 
  dplyr::mutate(VC_members = Genome, .keep = "unused")
```

There are two contigs contain BCoAT genes, each of them in a VC cluster that contains more than one member contigs,

```{r BCoAT_VCs}
vc_members %>% 
  DT::datatable(class = "white-space: nowrap")
```


```{r list-files-target, include=F}
projthis::proj_dir_info(path_target())
```
