---
title: "20221216-BCoAT_genome_compare"
title-block-banner: true
author:
  - name: Jinlong Ru
    orcid: 0000-0002-6757-6018
    email: jinlong.ru@gmail.com
    affiliations:
      - name: Institute of Virology, Helmholtz Centre Munich â€“ German Research Center for Environmental Health, 85764 Neuherberg, Germany
      - neme: Chair of Prevention of Microbial Diseases, School of Life Sciences Weihenstephan, Technical University of Munich, 85354 Freising, Germany
date: 2022-12-16
toc: true
format: gfm
params:
  name: "20221216-BCoAT_genome_compare"
---

**Updated: `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'CET')` CET.**

```{r setup, message=FALSE, include=FALSE}
here::i_am(paste0(params$name, ".qmd"), uuid = "03b88c18-6b1b-4c33-8dee-ef99417d2891")
projthis::proj_create_dir_target(params$name, clean = FALSE)
path_target <- projthis::proj_path_target(params$name)
path_source <- projthis::proj_path_source(params$name)

library(here)
library(tidyverse)
library(gggenomes)
library(p0064gvhd)
```

# Visualize using `gggenomes`

```{r BAF3}
dpath <- path_source("00-rawdata/gggenomes/BAF3")
fgbk_BAF3 <- here(dpath, "BAF3.gbk")
fgbk_BAF3_bac <- here(dpath, "NZ_QUGI01000002.1.gbk")
fgbks_BAF3 <- c(fgbk_BAF3, fgbk_BAF3_bac)
ftblastx <- here(dpath, "12.easyfig.out")

df_tblastx <- data.table::fread(ftblastx, header = FALSE, col.names = c("seq_id", "seq_id2", "pident", "length", "mismatch", "gapopen", "start", "end", "start2", "end2", "evalue", "bitscore")) %>% 
  mutate(seq_id = "VC-1", seq_id2 = "Pseudoflavonifractor sp. AF19-9AC") %>% 
  dplyr::select(c(seq_id, start, end, seq_id2, start2, end2, evalue, bitscore, pident)) %>% 
  dplyr::filter(evalue < 0.001, bitscore>50)

# fgbk_BAM6 <- here(dpath, "BAM6_00020/BAM6.gbk")

# read in gbk files for gggenomes
df_genes <- read_feats(fgbks_BAF3) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "NZ_QUGI01000002", "Pseudoflavonifractor sp. AF19-9AC", seq_id)) %>% 
  dplyr::mutate(type = ifelse(type == "HIT", "CDS", type)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "BAF3", "VC-1", seq_id)) %>% 
  dplyr::mutate(hit = ifelse(locus_tag %in% c("BAF3_00080", "DWX58_RS03030"), "yes", "no"))
  

df_seqs <- read_seqs(fgbks_BAF3) %>% 
  dplyr::select(c(seq_id, length)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "NZ_QUGI01000002", "Pseudoflavonifractor sp. AF19-9AC", seq_id)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "BAF3", "VC-1", seq_id))
# 
df_seqs[["start"]] <- c(62500, 32000)
df_seqs[["end"]] <- c(74000, 46000)
df_seqs[["length"]] <- df_seqs[["end"]] - df_seqs[["start"]]

p <- gggenomes(seqs = df_seqs, genes = df_genes, links = df_tblastx) +
  geom_seq(size=1) +
  geom_bin_label() +
  geom_gene(aes(fill = hit), size=2, show.legend = F ) +
  scale_fill_manual(values = c("#ffffbf", "#fc8d59")) +
  ggnewscale::new_scale("fill") +
  geom_link(aes(fill = pident), color="white", alpha=0.95, offset = 0.04) +
  # fill color gradient from white to red
  scale_fill_gradient(low = "white", high = "#91bfdb") +
  geom_gene_label(aes(label = product), nudge_y = 0.06, angle = 30, size = 3, check_overlap = T) +
  # change legend title to "identity (%)"
  labs(fill = "identity (%)")

ggsave(path_target("synteny_BAF3.svg"), p, width = 18, height = 5, limitsize = F)
```


```{r BAM6}
dpath <- path_source("00-rawdata/gggenomes/BAM6")
fgbk_BAF3 <- here(dpath, "BAM6.gbk")
fgbk_BAF3_bac <- here(dpath, "NZ_JAKNTI010000022.1.gbk")
fgbks_BAF3 <- c(fgbk_BAF3, fgbk_BAF3_bac)
ftblastx <- here(dpath, "12.easyfig.out")

df_tblastx <- data.table::fread(ftblastx, header = FALSE, col.names = c("seq_id", "seq_id2", "pident", "length", "mismatch", "gapopen", "start", "end", "start2", "end2", "evalue", "bitscore")) %>% 
  mutate(seq_id = "VC-2", seq_id2 = "Lawsonibacter sp. OA9") %>% 
  dplyr::select(c(seq_id, start, end, seq_id2, start2, end2, evalue, bitscore, pident)) %>% 
  dplyr::filter(evalue < 0.001, bitscore>30)

# fgbk_BAM6 <- here(dpath, "BAM6_00020/BAM6.gbk")

# read in gbk files for gggenomes
df_genes <- read_feats(fgbks_BAF3) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "NZ_JAKNTI010000022", "Lawsonibacter sp. OA9", seq_id)) %>% 
  dplyr::mutate(type = ifelse(type == "HIT", "CDS", type)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "BAM6", "VC-2", seq_id)) %>% 
  dplyr::mutate(hit = ifelse(locus_tag %in% c("BAM6_00020", "MCG44_RS03045"), "yes", "no"))
  

df_seqs <- read_seqs(fgbks_BAF3) %>% 
  dplyr::select(c(seq_id, length)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "NZ_JAKNTI010000022", "Lawsonibacter sp. OA9", seq_id)) %>% 
  dplyr::mutate(seq_id = ifelse(seq_id == "BAM6", "VC-2", seq_id))
# 
df_seqs[["start"]] <- c(15000, 420000)
df_seqs[["end"]] <- c(25000, 430000)
df_seqs[["length"]] <- df_seqs[["end"]] - df_seqs[["start"]]

# df_tblastx2 <- df_tblastx %>% 
  # dplyr::rename(start2 = end2, end2 = start2)

p <- gggenomes(seqs = df_seqs, genes = df_genes, links = df_tblastx) +
  geom_seq(size=1) +
  geom_bin_label() +
  geom_gene(aes(fill = hit), size=3, show.legend = F ) +
  scale_fill_manual(values = c("#ffffbf", "#fc8d59")) +
  ggnewscale::new_scale("fill") +
  geom_link(aes(fill = pident), color="white", alpha=0.95, offset = 0.06) +
  # fill color gradient from white to red
  scale_fill_gradient(low = "white", high = "#91bfdb") +
  geom_gene_label(aes(label = product), nudge_y = 0.08, angle = 20, size = 2.5, check_overlap = T) +
  # gggenomes::flip(2) +
  # change legend title to "identity (%)"
  labs(fill = "identity (%)")

ggsave(path_target("synteny_BAM6.svg"), p, width = 20, height = 4, limitsize = F)
```


------------------------------------------------------------------------

**Updated: 2022-12-16**

Two Viral BCoAT protein sequences were searched against NCBI NR database. The top protein hit in bacterial reference genome was selected as the best target, and the bacterial genome was compared with viral genome as show below.

### BCoAT in viral contig BAF3_V\_\_NODE_17

Bacterial `Pseudoflavonifractor sp. AF19-9AC` contains BCoAT that most similar to viral contig `BAF3_V__NODE_17` encoded BCoAT gene, so we compared BCoAT region of their genomes, as show below. BCoAT gene was colored in red, and was connected with lines according to identity by BLASTn alignment. The high identity in BCoAT gene and no identity in surrounding genes indicate the viral-encoded BCoAT gene was from bacterial genome. Upper part is bacterial genome, and lower part is viral contig.

![Figure 1. BCoAT in Pseudoflavonifractor sp. AF19-9AC and BAF3_V\_\_NODE_17](data/00-rawdata/easyfig/BAF3.svg)

### BCoAT in viral contig BAM6_V\_\_NODE_3

Bacterial `Lawsonibacter sp. OA9` contains BCoAT that most similar to viral contig `BAM6_V__NODE_3` encoded BCoAT gene, so we compared BCoAT region of their genomes, as show below. BCoAT gene was colored in red, and was connected with lines according to identity by BLASTn alignment. The high identity in BCoAT gene and no identity in surrounding genes indicate the viral-encoded BCoAT gene was from bacterial genome. Upper part is bacterial genome, and lower part is viral contig.

![Figure 2. BCoAT in Lawsonibacter sp. OA9 and BAM6_V\_\_NODE_3](data/00-rawdata/easyfig/BAM6.svg)

```{r list-files-target, message=FALSE, include=FALSE}
projthis::proj_dir_info(path_target())
```
