---
title: "21-preprocess"
title-block-banner: true
author:
  - name: Jinling Xue
    orcid: 0000-0002-8191-4140
toc: true
code-fold: true
format: 
  html:
    embed-resources: true
    smooth-scroll: true
bibliography: refs.bib
params:
  name: "21-preprocess"
---

Updated: `r format(Sys.time(), '%Y-%m-%d %H:%M:%S', tz = 'CET')` CET.


```{r setup}
#| message: false
#| include: false
here::i_am(paste0(params$name, ".qmd"), uuid = "c1ebd54f-0c77-4665-a447-117cd32ca04d")
projthis::proj_create_dir_target(params$name, clean = FALSE)
path_target <- projthis::proj_path_target(params$name)
path_source <- projthis::proj_path_source(params$name)
path_data <- path_source("00-raw", params$name)
dir.create(path_data, recursive = T)
```

The purpose of this document is ...

```{r packages}
library(here)
library(tidyverse)
library(data.table)
library(phyloseq)
library(gridExtra)
library(reshape2)
library(ggpubr)
library(speedyseq)
library(gvhdvirome)
```

## Tasks

The first task is ...

### Create phyloseq objects

```{r}


dis_stat_vir <- c("Pre-Tx", "d0", "d7", "d14", "d28", "aGvHD")
dis_stat_16s <- c("Pre-Tx", "d0", "d7", "d14", "d28", "aGvHD", "Follow Up")
dis_stat_18s <- c("Pre-TX", "d0", "d7", "d14", "d28", "aGvHD", "Follow_up")

# --------- File path --------------
fpath_metadata = here("data/processed/metadata.tsv")
fpath_16s <- here("data/collab/2021-10-20_MOFA-additional-notes_email/TaxaSubset_Genus_V1V3.xlsx")
fpath_ITS <- here("data/collab/2021-10-20_MOFA-additional-notes_email/TaxaSubset_Genus_ITS.xlsx")
# fpath_vabundance = here("data/hms/virmerge/files/abundance/abundance_contigs_count.tsv")
fpath_vabundance = here("vabundance.tsv")
fpath_vOTUs = here("data/hms/virmerge/files/clusters/vOTUs.tsv")
fpath_vir_merged = here("data/hms/virmerge/files/merge3.tsv")


# ---------- metadata -------------
metadata <- read_tsv(fpath_metadata)
pseq_meta <- metadata %>% 
    column_to_rownames("Project_ID") %>% 
    sample_data()

# ------------ 16S -----------------
amp16s_counts <- openxlsx::read.xlsx(fpath_16s, sheet = "Counts") %>% 
    column_to_rownames("OTU") %>% 
    as.matrix() %>% 
    otu_table(taxa_are_rows = TRUE)

amp16s_taxa <- openxlsx::read.xlsx(fpath_16s, sheet = "Taxonomy") %>% 
    column_to_rownames("OTU") %>% 
    as.matrix() %>% 
    tax_table()


pseq_16s <- phyloseq(amp16s_counts, amp16s_taxa, pseq_meta)
save(pseq_16s, file=here("data/processed/pseq_16s.RData"))


# ---------- ITS ---------------

ampITS_counts <- openxlsx::read.xlsx(fpath_ITS, sheet = "Counts") %>% 
    column_to_rownames("OTU") %>% 
    as.matrix() %>% 
    otu_table(taxa_are_rows = TRUE)

ampITS_taxa <- openxlsx::read.xlsx(fpath_ITS, sheet = "Taxonomy") %>% 
    column_to_rownames("OTU") %>% 
    as.matrix() %>% 
    tax_table()

pseq_ITS <- phyloseq(ampITS_counts, ampITS_taxa, pseq_meta)
save(pseq_ITS, file=here("data/processed/pseq_ITS.RData"))

# ---------- Virome ------------
# votu <- merge_vOTU_abundance(fpath_vabundance, fpath_vOTUs) %>%
#     column_to_rownames("contig_id_raw") %>%
#     dplyr::select(colnames(.)[colnames(.) %in% metadata$vir_seq_sample_id])

votu <- fread(fpath_vabundance) %>% 
  column_to_rownames("Contig") %>% 
  dplyr::select(colnames(.)[colnames(.) %in% metadata$vir_seq_sample_id])

vmeta <- metadata %>% 
    filter(metadata$vir_seq_sample_id %in% colnames(votu)) %>% 
    column_to_rownames("Project_ID") %>% 
    sample_data()
    
votu <- votu %>% 
    rename(set_names(vmeta$vir_seq_sample_id, rownames(vmeta))) %>% 
    as.matrix() %>% 
    otu_table(taxa_are_rows = TRUE)

vtax <- extract_viral_taxa(fpath_vir_merged) %>%
    column_to_rownames("contig_id_raw") %>% 
    as.matrix() %>% 
    tax_table()

# Create phyloseq object
# pseq_vir <- phyloseq(votu, vtax, pseq_meta)
pseq_vir <- phyloseq(votu, pseq_meta)
save(pseq_vir, file=here("data/processed/pseq_vir.RData"))
```


### Merge patients

**Updated: 2023-01-29**

```{r data path}
fin_ra <- here(path_data, "virome-RA.xlsx")
fout_patient <- path_target("virome_RA_mergeby_patient.xlsx")
fout_antibiotics <- path_target("virome_RA_mergeby_antibiotics.xlsx")
fin_af <- here(path_data, "virome-AF.xlsx")
fout_af <- path_target("virome_RA_mergeby_antifungal.xlsx")
```



```{r by patient}
df_patient2sample <- openxlsx::read.xlsx(fin_ra, sheet = "Group-1") %>% 
  setnames(colnames(.), c("sample_id", "project_id", "with_gvhd", "no_gvhd")) %>% 
  mutate(patient_id = str_replace_all(project_id, "-[0-9]+$", "")) %>% 
  dplyr::select(c("sample_id", "patient_id"))

df_relab_patient <- openxlsx::read.xlsx(fin_ra, sheet = "GVHD") %>% 
  dplyr::filter(!is.na(contig_id)) %>% 
  column_to_rownames("contig_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("sample_id") %>% 
  dplyr::left_join(df_patient2sample, by = "sample_id") %>% 
  dplyr::select(-"sample_id") %>% 
  group_by(patient_id) %>%
  summarise_all(sum) %>%
  column_to_rownames("patient_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("feature_id") %>% 
  mutate(feature_id = str_replace(feature_id, "vout.", "votu_"))
  
openxlsx::write.xlsx(df_relab_patient, fout_patient)
```

```{r by antibiotics}
df_abx2sample <- openxlsx::read.xlsx(fin, sheet = "Group-2") %>% 
  setnames(colnames(.), c("sample_id", "project_id", "with_antibiotics")) %>% 
  mutate(patient_id = str_replace_all(project_id, "-[0-9]+$", "")) %>% 
  mutate(patient_id = paste0(patient_id, "_", with_antibiotics)) %>% 
  dplyr::select(c("sample_id", "patient_id"))

df <- openxlsx::read.xlsx(fin, sheet = "ABX") %>% 
  dplyr::filter(!is.na(contig_id)) %>% 
  column_to_rownames("contig_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("sample_id") %>% 
  dplyr::left_join(df_abx2sample, by = "sample_id")

df2 <- df %>% 
  dplyr::filter(!is.na(patient_id)) %>% 
  dplyr::select(-"sample_id") %>% 
  group_by(patient_id) %>%
  summarise_all(sum) %>% 
  column_to_rownames("patient_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("feature_id") %>% 
  mutate(feature_id = str_replace(feature_id, "votu", "votu_"))

openxlsx::write.xlsx(df2, fout_antibiotics)
```

```{r by anti-fungal}
dfaf_meta <- openxlsx::read.xlsx(fin_af, sheet = "Group3") %>% 
  setnames(colnames(.), c("sample_id", "project_id", "hospital", "with_af")) %>% 
  mutate(patient_id = str_replace_all(project_id, "-[0-9]+$", "")) %>% 
  mutate(patient_id = paste0(patient_id, "_", with_af)) %>% 
  dplyr::select(c("sample_id", "patient_id"))

dfaf_assay <- openxlsx::read.xlsx(fin_af, sheet = "AF") %>% 
  dplyr::filter(!is.na(contig_id)) %>% 
  column_to_rownames("contig_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("sample_id") %>% 
  dplyr::left_join(dfaf_meta, by = "sample_id")

dfaf_assay2 <- dfaf_assay %>% 
  # dplyr::select(c(sample_id, patient_id))
  dplyr::filter(!is.na(patient_id)) %>% 
  dplyr::select(-"sample_id") %>% 
  group_by(patient_id) %>%
  summarise_all(sum) %>% 
  column_to_rownames("patient_id") %>% 
  t() %>% 
  data.frame() %>% 
  rownames_to_column("feature_id")

openxlsx::write.xlsx(dfaf_assay2, fout_af)
```


## Files written

These files have been written to the target directory, ```r paste0("data/", params$name)```:

```{r list-files-target}
projthis::proj_dir_info(path_target())
```
